subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/GroovyApplicationProjectInitDescriptor.java

```diff
@@ -27,8 +30,14 @@ public GroovyApplicationProjectInitDescriptor(TemplateOperationFactory templateO
     @Override
     protected void configureBuildScript(BuildScriptBuilder buildScriptBuilder) {
         buildScriptBuilder
-            .plugin("Apply the application plugin to add support for building an application", "application")
-            .configuration("Define the main class for the application", "mainClassName = 'App'");
+            .plugin(
+                "Apply the application plugin to add support for building an application",
+                "application")
+            .configuration(
+                "Define the main class for the application",
+                ImmutableMap.of(
+                    BuildInitBuildScriptDsl.GROOVY, "mainClassName = 'App'",
```

Comment:
❌  It would be nice to introduce some structure here so multiple configuration items could be squashed together (monoids, monoids everywhere) to produce  declarative style configuration blocks instead of an imperative sequence of assignments, in other words, it would be nice if, given multiple configuration items involving the `application` extension, the resulting Kotlin build script looked like:

```kotlin
application {
   mainClassName = "App"
   applicationName = "My App"
}
```

Instead of:

```kotlin
application.mainClassName = "App"
application.applicationName = "My App"
```

The other cool thing about introducing structure for the configuration snippets instead of using strings: clients of the API (the `*InitDescriptor`s in this case) don't have to care about the scripting language details:

```java
    .configuration(
        "Define the main class for the application",
        ConventionPropertyAssignment.of("application", "mainClassName", Constant.of("App"))
```

All scripting language syntax concerns can then be handled in a more cohesive fashion by language specific _configuration snippet printers_ instead of being sprinkled over the code base.

As a bonus, such configuration building expressions can then be more easily extracted to reduce duplication in the code base.

Reply:
Agreed. But we don't need this for this PR and the `init` plugin. This is definitely something we should do if we expand the scope of this `BuildScriptBuilder`, in another story.

Reply:
What do you mean by "we don't need this for this PR and the `init`plugin"?

I'd like to insist that it would be a much cleaner solution than spreading Kotlin and Groovy snippets all over.

Reply:
After contemplating this a bit I agree :) I'll come up with something tomorrow